test/runtests.jl

```diff
@@ -1,82 +1,120 @@
+                        rand($T) - $T(0.5)
+                    else
+                        rand($T)
+                    end
+                    # We're happy with types with the correct promotion behavior, e.g.
+                    # it's fine to return `1` as a derivative despite input being `Float64`.
+                    @test one($T) * $deriv isa $T
+                    @test $deriv ≈ finitediff($M.$f, goo) rtol=1e-2 atol=1e-3
+                    # test for 2pi functions
+                    if $(f === :mod2pi)
+                        goo = 4 * pi
+                        @test NaN === $deriv
+                    end
                 end
             end
-        end
-    elseif arity == 2
-        @test DiffRules.hasdiffrule(M, f, 2)
-        derivs = DiffRules.diffrule(M, f, :foo, :bar)
-        @eval begin
-            let
-                if "mod" == string($M.$f)
-                    foo, bar = rand() + 13, rand() + 5 # make sure x/y is not integer
-                else
-                    foo, bar = rand(1:10), rand()
-                end
-                dx, dy = $(derivs[1]), $(derivs[2])
-                if !(isnan(dx))
-                    @test isapprox(dx, finitediff(z -> $M.$f(z, bar), float(foo)), rtol=0.05)
-                end
-                if !(isnan(dy))
-                    @test isapprox(dy, finitediff(z -> $M.$f(foo, z), bar), rtol=0.05)
+        elseif arity == 2
+            @test DiffRules.hasdiffrule(M, f, 2)
+            derivs = DiffRules.diffrule(M, f, :foo, :bar)
+            @eval begin
+                let
+                    foo, bar = if $(f === :mod)
+                        rand($T) + 13, rand($T) + 5 # make sure x/y is not integer
+                    elseif $(f === :polygamma)
+                        rand(1:10), rand($T) # only supports integers as first arguments
+                    elseif $(f in (:bessely, :besselyx))
+                        # avoid singularities with finite differencing
+                        rand($T), rand($T) + $T(0.5)
+                    elseif $(f === :log)
+                        # avoid singularities with finite differencing
+                        rand($T) + $T(1.5), rand($T)
+                    elseif $(f === :^)
+                        # avoid singularities with finite differencing
+                        rand($T) + $T(0.5), rand($T)
+                    else
+                        rand($T), rand($T)
+                    end
+                    dx, dy = $(derivs[1]), $(derivs[2])
+                    if !(isnan(dx))
+                        @test dx ≈ finitediff(z -> $M.$f(z, bar), foo) rtol=1e-2 atol=1e-3
+
+                        # Check type, if applicable.
+                        if foo isa AbstractFloat
+                            if dx isa Complex
+                                @test dx isa Complex{$T}
+                            else
+                                @test dx isa $T
+                            end
```

Comment:
We don't have to be worried about promotions here?
```suggestion
                            @test promote_type(typeof(real(dx)), $T) === $T
```